crates/mpz-garble-core/src/evaluator.rs

```diff
@@ -122,111 +115,204 @@ impl Evaluator {
                 } => {
-                    let x = labels[node_x.id()].expect("feed should be initialized");
-                    let y = labels[node_y.id()].expect("feed should be initialized");
-                    labels[node_z.id()] = Some(x ^ y);
+                    let x = self.labels[node_x.id()];
+                    let y = self.labels[node_y.id()];
+                    self.labels[node_z.id()] = x ^ y;
                 }
                 Gate::And {
                     x: node_x,
                     y: node_y,
                     z: node_z,
                 } => {
-                    if let Some(encrypted_gate) = encrypted_gates.next() {
-                        if let Some(hasher) = &mut self.hasher {
-                            hasher.update(&encrypted_gate.to_bytes());
-                        }
-
-                        let x = labels[node_x.id()].expect("feed should be initialized");
-                        let y = labels[node_y.id()].expect("feed should be initialized");
-                        let z = and_gate(self.cipher, &x, &y, encrypted_gate, self.gid);
-                        labels[node_z.id()] = Some(z);
-                        self.gid += 2;
-                    } else {
-                        // We ran out of encrypted gates, so we return until we get more
+                    let x = self.labels[node_x.id()];
+                    let y = self.labels[node_y.id()];
+                    let z = and_gate(self.cipher, &x, &y, &encrypted_gate, self.gid);
+                    self.labels[node_z.id()] = z;
+
+                    self.gid += 2;
+                    self.counter += 1;
+
+                    if let Some(hasher) = &mut self.hasher {
+                        hasher.update(&encrypted_gate.to_bytes());
+                    }
+
+                    // If we have more AND gates to evaluate, return.
+                    if self.wants_gates() {
                         return;
                     }
                 }
+                Gate::Inv {
+                    x: node_x,
+                    z: node_z,
+                } => {
+                    let x = self.labels[node_x.id()];
+                    self.labels[node_z.id()] = x;
+                }
             }
-            self.pos += 1;
         }
 
         self.complete = true;
     }
 
-    /// Returns whether the evaluator has finished evaluating the circuit.
-    pub fn is_complete(&self) -> bool {
-        self.complete
-    }
-
-    /// Returns the active encoded outputs of the circuit.
-    pub fn outputs(&self) -> Result<Vec<EncodedValue<state::Active>>, EvaluatorError> {
-        if !self.is_complete() {
+    /// Returns the encoded outputs of the circuit.
+    pub fn finish(mut self) -> Result<EvaluatorOutput, EvaluatorError> {
+        if self.wants_gates() {
             return Err(EvaluatorError::NotFinished);
         }
 
-        Ok(self
-            .circ
-            .outputs()
+        // Evaluate the remaining "free" gates.
```

Comment:
This is a special case which applies only if there are 0 AND gates, in all other cases if wants_gates() == false, it means that complete == true. Making this more explicit in the comment
```suggestion
        // If there were 0 AND gates in the circuit, we need to evaluate the "free" gates now.
```

Reply:
I think `self.wants_gates() == false` only means that the circuit does not contain any more AND gates. But it is still possible that the tail contains non-evaluated XOR, INV gates, which need one more call to `next` so that `self.complete == true`.

Reply:
My suggested comment is still valid, ill explain on the call today.

Reply:
That's not how I see the code works, I'll explain on the call today.